lib/rouge/lexer.rb

```diff
@@ -501,12 +521,47 @@ def self.detect?(text)
   end
 
   module Lexers
+    ProxyLexer = Struct.new :path, :name, :tag, :aliases, :filenames, :mimetypes
+
+    @_loaded_files = {}
     @_loaded_lexers = {}
+    @_loaded_proxies = {}
+    @_source_files = []
+
+    def self.add_lexer(class_name)
+      return if @_source_files.empty?
+      @_loaded_files[class_name] = @_source_files.last
+    end
 
     def self.load_lexer(relpath)
       return if @_loaded_lexers.key?(relpath)
       @_loaded_lexers[relpath] = true
+      @_source_files.push relpath
       load File.join(__dir__, 'lexers', relpath)
+      @_source_files.pop
```

Comment:
Why push `relpath` into `@_source_files` only to be taken out here..?

Reply:
This definitely needs a comment to explain it (not least because I will forget why it did this in six months).

So, the problem we have is that we need to build an index of lexer classes with the filenames in which the class is defined. Ruby doesn't provide any facility to do this because a class can be defined across multiple files so it doesn't make sense to know where it was 'defined'.

While it's theoretically possible to define a class across multiple files, that's not how people define lexers with Rouge (if they do, this caching approach will break). To build the index, I save the name of the file about to be loaded to `@_source_files`, then load the file, taking advantage of `Class#inherited` to assign what is effectively a callback to `Lexer.inherited` that will call `Lexers.add_lexer` when a lexer class is loaded. When we've finished loading, I remove the filename from `@_source_files`.

That's probably pretty straightforward and it is pretty straightforward in the basic case. The problem is when the lexer that we're loading loads a lexer. If we don't update the filename, we'll associate the filename of the calling class rather than the class itself. If we replace the filename, then when we finish loading the included lexer. In this case, we'll have the wrong filename associated with the calling lexer. The solution is to use a FILO queue that we can push and pop filenames from.

Phew. I sort of hope this problem is inherently difficult and I didn't spend a good chunk of today just not seeing the simple solution!

Reply:
:open_mouth: This seems to be very complicated (and therefore prone to bugs).

Most Ruby  libraries just depend on `Module#autoload` to associate a filename to a class and load it into memory as required. Perhaps Rouge's architecture could be tweaked to depend on that..?

Reply:
I think if you look at the way `load_lexer` is currently defined, you'll find that we already rely on mapping lexer names to filenames. If we say that user-contributed lexers (ones not in this repository) don't support lazy loading, does this make the situation easier? We could make use of `const_missing`...

Reply:
@jneen I don't know if the new approach makes addressing this comment unnecessary but for what it's worth, `load_lexer` does use filenames but there's no simple way to know after loading a file what classes were defined in it. You could do a comparison with the ObjectSpace before and after but that didn't seem like a great idea. My solution was to take advantage of the fact that all lexers inherit from `Rouge::Lexer` and use the `Rouge::Lexer.inherited` method.

Unfortunately, the only data that Ruby passes to the `inherited` method is the name of the inheriting class which, by itself, isn't that helpful. To provide `Rouge::Lexer.inherited` with the name of the file at the time it's called, we need to store the value somewhere else. My initial approach was to add `@_source_files` to `Rouge::Lexers` and access it directly but that's kind of gross. The new approach is still kind of gross but at least looks nicer. Since I'm defining the `Rouge::Lexer.inherited` method in a block in the `generate:cache` Rake task, I can put the array of loading files into scope by assigning the value to a local variable that's accessible from the block (I actually put an instance of `Rouge::Tasks::ProxyBuilder` into scope and make the list of files an instance variable but that's really just an implementation detail).